plugin/cassandra/plugin.go

```diff
@@ -0,0 +1,666 @@
+		fail = true
+	} else {
+		fmt.Printf("@G{\u2713 cassandra_keyspace}      @C{%s}\n", s)
+	}
+
+	s, err = endpoint.StringValueDefault("cassandra_bindir", "")
+	if err != nil {
+		fmt.Printf("@R{\u2717 cassandra_bindir        %s}\n", err)
+		fail = true
+	} else if s == "" {
+		fmt.Printf("@G{\u2713 cassandra_bindir}        using default @C{%s}\n", DefaultBinDir)
+	} else {
+		fmt.Printf("@G{\u2713 cassandra_bindir}        @C{%s}\n", s)
+	}
+
+	s, err = endpoint.StringValueDefault("cassandra_datadir", "")
+	if err != nil {
+		fmt.Printf("@R{\u2717 cassandra_datadir       %s}\n", err)
+		fail = true
+	} else if s == "" {
+		fmt.Printf("@G{\u2713 cassandra_datadir}       using default @C{%s}\n", DefaultDataDir)
+	} else {
+		fmt.Printf("@G{\u2713 cassandra_datadir}       @C{%s}\n", s)
+	}
+
+	s, err = endpoint.StringValueDefault("cassandra_tar", "")
+	if err != nil {
+		fmt.Printf("@R{\u2717 cassandra_tar           %s}\n", err)
+		fail = true
+	} else if s == "" {
+		fmt.Printf("@G{\u2713 cassandra_tar}           using default @C{%s}\n", DefaultTar)
+	} else {
+		fmt.Printf("@G{\u2713 cassandra_tar}           @C{%s}\n", s)
+	}
+
+	if fail {
+		return fmt.Errorf("cassandra: invalid configuration")
+	}
+	return nil
+}
+
+// Backup one cassandra keyspace
+func (p CassandraPlugin) Backup(endpoint plugin.ShieldEndpoint) error {
+	cassandra, err := cassandraInfo(endpoint)
+	if err != nil {
+		return err
+	}
+
+	plugin.DEBUG("Cleaning any stale '%s' snapshot", SnapshotName)
+	cmd := fmt.Sprintf("%s/nodetool clearsnapshot -t %s \"%s\"", cassandra.BinDir, SnapshotName, cassandra.Keyspace)
+	plugin.DEBUG("Executing: `%s`", cmd)
+	err = plugin.Exec(cmd, plugin.STDIN)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "@R{\u2717 Clean any stale snapshot}\n")
+		return err
+	}
+	fmt.Fprintf(os.Stderr, "@G{\u2713 Clean any stale snapshot}\n")
+
+	defer func() {
+		plugin.DEBUG("Clearing snapshot '%s'", SnapshotName)
+		cmd := fmt.Sprintf("%s/nodetool clearsnapshot -t %s \"%s\"", cassandra.BinDir, SnapshotName, cassandra.Keyspace)
+		plugin.DEBUG("Executing: `%s`", cmd)
+		err := plugin.Exec(cmd, plugin.STDIN)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "@R{\u2717 Clean snapshot}\n")
+			return
+		}
+		fmt.Fprintf(os.Stderr, "@G{\u2713 Clean snapshot}\n")
+	}()
+
+	plugin.DEBUG("Creating a new '%s' snapshot", SnapshotName)
+	cmd = fmt.Sprintf("%s/nodetool snapshot -t %s \"%s\"", cassandra.BinDir, SnapshotName, cassandra.Keyspace)
+	plugin.DEBUG("Executing: `%s`", cmd)
+	err = plugin.Exec(cmd, plugin.STDIN)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "@R{\u2717 Create new snapshot}\n")
+		return err
+	}
+	fmt.Fprintf(os.Stderr, "@G{\u2713 Create new snapshot}\n")
+
+	// Here we need to copy the snapshots/shield-backup directories into a
+	// {keyspace}/{tablename} structure that we'll temporarily put in
+	// /var/vcap/store/shield/cassandra. Then we can tar it all and stream
+	// that to stdout.
+
+	baseDir := "/var/vcap/store/shield/cassandra"
```

Comment:
I would prefer not to use `store` for this; ideally, golang's `archive/tar` library will allow you to stream the archive, with arbitrary paths, without touching the filesystem (which can be very important if you are backing up a thing and don't want to enforce 50% disk usage max).

Reply:
Interesting approach. But the comment is wrong and needs to be updated.

At backup time, we actually hard-link the created immutable snapshot files into a suitable directory structure, that we delete once the tar file is streamed. Cassandra generates the snapshot files only in subdirectories of its own data dir. So, it's already in `/var/vcap/store` and there's nothing we can do about that.

At restore time, the `sstableloader` tool also needs the whole tar archive to be extracted somewhere before it starts working. Here we've chosen to put that on the persistent disk. I'm raising this point here because we could do differently and put that on the ephemeral disk in `/var/vcap/data`. For the time being, we just thought that if we need the extra space at backup time, then we can use it at restore time.

All in all, our current approach would be to document (here in this plugin and in the Cassandra Bosh release) the need for extra space in `/var/vcap/store`. But don't hesitate to suggest any other way of seeing things!

Reply:
Ah, so it's an in-built limitation of the actual backup strategy.  As long as we document that and explain it, I'm fine with it (although I do hate requiring all that extra storage...)